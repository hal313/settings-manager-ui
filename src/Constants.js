const ATTRIBUTE_PREFIX = 'data-setting';

const Constants = {
    ATTRIBUTE_PREFIX: ATTRIBUTE_PREFIX,
    ATTRIBUTE_NAME: `${ATTRIBUTE_PREFIX}-name`,
    ATTRIBUTE_TYPE: `${ATTRIBUTE_PREFIX}-type`,
    ATTRIBUTE_CONTAINER_ELEMENT: `${ATTRIBUTE_PREFIX}-container`
};

export {Constants};