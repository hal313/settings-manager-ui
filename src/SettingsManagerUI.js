/* global document */

// Build User: ${build.user}
// Version: ${build.version}
// Build Date: ${build.date}


// NOTE: Run specific test:               npx jest --color --testRegex '.*ObjectDefaultTypeHandler.spec.*' --watch --collectCoverage false
// NOTE: Run specific test with coverage: npx jest --color --testRegex '.*ObjectDefaultTypeHandler.spec.*' --collectCoverageFrom='**/*ObjectDefaultTypeHandler*' --watch --coverage true


// npx rollup src/SettingsManagerUI.js --file dist/test.js --format iife --name "Bund" --globals 'template-manager:TemplateManager'
// npx rollup src/SettingsManagerUI.js --file dist/test.js --format iife --name "Bund" --globals 'template-manager:TemplateManager' ## watch?
// npx rollup src/SettingsManagerUI.js --file dist/test.js --format umd --name "SettingsManagerUI" --globals 'template-manager:TemplateManager'
// tsc -p jsconfig.json --outDir build

// NEXT:
// [string|boolean|number|object]arrays
// rollup

// TODO: getTypeFromValue should have configuration for collection type detection
// collection:object
// TODO: If the type changes (collection:object), the settingModifer will use the existing type handler instead of the new type handler. how to handle this?
// TODO: collection:object is not nestable because of the searches; figure out how to leverage the object:default lookups
// TODO: Consider adding [data-setting-value] to values?
// TODO: Test complex types (objects with object[array] children)
// TODO: Enable skipped tests


// TODO: Does applying decorators make getValue break if the decorated element is embedded? Can we disallow decorators to return other elements? CollectionObjectTypeHandler uses 'const valueElement = getChildSettingElements(childElement)[0];' to work around this; should we check for only 1 element? getOnlyChildSettingElement()?
// TODO: Apply flat-object-name during population
// TODO: Pass type, name, value, flatName through to decorators


// TODO: Add fallback handler (.*, default to string)? may require matching on type lookup?
// TODO: Handlers should type-check value parameter


// TODO: Support object/collection sorters
// TODO: Support value decorators
// TODO: Support sorter/value decorator/element decorator assignment delcaratively


// TODO: Create TypeHandlerManagerConfigurator -> use in tests and for settingsuiconfigurator
// TODO: Should typeHandler.setElement() should set attributes? NO! fix in object/collection classes not type if it exists, maybe even if it doesnt. But certainly name and possibly container?


// TODO: Add "remove" button for elements in playgound (can use a decorator!)
// TODO: Add hotkeys for set/update buttons in playground
// TODO: Why does the playground page keep flashing? Is it because of imports?


// TODO: Use template manager
// TODO: Flat mode / Hierarchical mode / add data-setting-manager-flat-name="some.nested.value" to all elements
// TODO: Callbacks for type overrides (pass name/value or element)


// TODO: Get jest plugin working
// TODO: Debug jest in ide


// TODO: Fix gruntfile/move to gulp/add clean target?


// TODO: Build CLI - use JSDOM(htmlstring) to help?
// TODO: Code examples for grunt/gulp tasks


// TODO: Support ng style conversions (someName === some-name)


// TODO: Can the xxxTypeHandler objects be singletons?
// TODO: Might be possible to optimize (less iteration? combine both childElements.forEach loops?) (object/collection handlers)
// TODO: debug flag
// TODO: jsdoc
// TODO: README
// TODO: How to handle when putSettingsIntoUI() is called and some settings dont exist in the DOM? We should track this, right? Keep a copy of settings and merge with UI on getSettingsFromUI?
// TODO: Add updateSettingsIntoElement(settings, element) => Update existing settings only by changing value/checked/etc
// TODO: Support renderer running the processChain on values at render time
// TODO: Support renderer template resolver maps
// TODO: Should populateObjectElement() use own properties? or all? or a selector to decide which properties to use?
// TODO: Figure out a way to not need uniqueClass on lookups (getObjectValueFromElement())
// TODO: Different template managers for arrayobject, renderer, other templates (map of templatemanagers?)
// TODO: Support a way to use custom renderers/templates/resolvers for renders
// TODO: Type override map (in case numbers are stored as strings, for example)
// TODO: Extensible renderer functionality (use external renderer, based on type/value/name)
// TODO: Primitive array types (string, number, boolean)?
// TODO: constants for types (string, number, boolean, object, objectarray)
// TODO: Ignore read/ignore write processors?
// TODO: Add support for processors at the object level (apply to all members; how to support ordering if a sub-element has processors, too? pre/post?)
// TODO: Reorg constants
// TODO: What if populateXXX with a non-boolean is passed to a checkbox?
// TODO: Make a jquery plugin
// TODO: Allow for non-flat options {debug: {enabled: true, only-on-change: true}}
// TODO: In readme, talk about how this can be a declarative solution to options
// TODO: Allow types to be specified for each array value
// TODO: Get rid of data-setting-xxx and move to attributes (setting-xxx)
// TODO: Consider having the configurator pull in descriptors from JSON and load them (note: this might be challenging because of the need to dynamically import paths and import statements can be top level only)

import { getOneElement } from './util/getOneElement.js';
import { TypeDecoratorManager } from './util/TypeDecoratorManager.js';
import { TypeHandlerManager } from './util/TypeHandlerManager.js';
import { SettingsManagerUIConfigurator } from './util/SettingsManagerUIConfigurator.js';
import { SettingModifier } from './util/SettingModifier.js';


// OK, so this will suck with a lot of instances because they wont be garbage collected :(
const privates = {};
const getPrivate = (instance, name) => {
    if (!privates[instance]) {
        privates[instance] = {};
    }
    return privates[instance][name];
};
const setPrivate = (instance, name, value) => {
    if (!privates[instance]) {
        privates[instance] = {};
    }
    privates[instance][name] = value;
};

export class SettingsManagerUI {

    constructor() {

        let typeHandlerManager = new TypeHandlerManager();
        let typeDecoratorManager = new TypeDecoratorManager();
        let settingModifier = new SettingModifier(typeHandlerManager, typeDecoratorManager);

        setPrivate(this, 'typeHandlerManager', typeHandlerManager);
        setPrivate(this, 'typeDecoratorManager', typeDecoratorManager);
        setPrivate(this, 'settingModifier', settingModifier);

        // Configure this instance
        SettingsManagerUIConfigurator.configure(this);
    };

    addTypeHandler(typeHandler) {
        getPrivate(this, 'typeHandlerManager').addTypeHandler(typeHandler);
    };

    addTypeDecorator(type, decoratorFn) {
        getPrivate(this, 'typeDecoratorManager').addTypeDecorators(type, decoratorFn);
    };

    addTypeDecorators(type, decoratorFns) {
        this.addTypeDecorator(type, decoratorFns);
    };

    setDefaultHandler(type, typeHandler) {
        getPrivate(this, 'typeHandlerManager').setDefaultHandler(type, typeHandler);
    }

    setValueTo(value, target) {
        getPrivate(this, 'settingModifier').setValue(getOneElement(target), value);
    };

    getValueFrom(target) {
        return getPrivate(this, 'settingModifier').getValue(getOneElement(target));
    };

    setSettings(settings, target) {
        return this.setValueTo(settings, target);
    };

    getSettings(target) {
        return this.getValueFrom(target);
    };

}

// Place the version as a member in the function
SettingsManagerUI.version = '${build.version}';