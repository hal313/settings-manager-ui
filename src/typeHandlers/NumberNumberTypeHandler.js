import {isDefinedAndNotEmpty} from '../util/isDefinedAndNotEmpty.js';
import {createError} from '../util/createError.js';
import { isElement } from '../util/isElement.js';
import { isNumber } from '../util/isNumber.js';
import { TypeHandler } from './TypeHandler.js';

export class NumberNumberTypeHandler extends TypeHandler
{

    constructor() {
        super();
    }

    getType() {
        return NumberNumberTypeHandler.TYPE;
    }

    getValue(element) {
        return +element.value;
    }

    setValue(element, value) {
        if (!isElement(element)) {
            throw createError('The "element" parameter must be an Element');
        }

        element.value = +value;
    }

    createElement(name, value) {
        if (!isDefinedAndNotEmpty(name)) {
            throw createError('The "name" parameter must be specified');
        }

        let element = document.createElement('input');
        element.setAttribute('name', name);
        element.setAttribute('type', 'number');
        element.setAttribute('value', value);

        this.setValue(element, value);

        return element;
    }

};
NumberNumberTypeHandler.TYPE = 'number:number';