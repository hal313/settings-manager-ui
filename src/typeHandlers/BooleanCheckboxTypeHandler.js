import { isDefinedAndNotEmpty } from '../util/isDefinedAndNotEmpty.js';
import { createError } from '../util/createError.js';
import { Constants } from '../Constants.js';

export class BooleanCheckboxTypeHandler {

    constructor() {}

    getType() {
        return 'boolean:checkbox';
    }

    getValue(element) {
        // return !!element.value;
        return !!element.hasAttribute('checked');
    }

    setValue(element, value) {
        // element.value = !!value
        !!value ? element.setAttribute('checked', '') : element.removeAttribute('checked');
    }

    createElement(name, value) {
        if (!isDefinedAndNotEmpty(name)) {
            throw createError('The "name" parameter must be specified');
        }

        // return asElement(`<input name="${name}" ${Constants.ATTRIBUTE_NAME}="${name}" ${Constants.ATTRIBUTE_TYPE}="${this.getType()}" type="checkbox" ${!!value?'checked':''}/> ${name}`);
        let element = document.createElement('input');
        element.setAttribute('type', 'checkbox');
        element.setAttribute('name', name);
        if (!!value) {
            element.setAttribute('checked', '');
        }
        element.setAttribute(Constants.ATTRIBUTE_NAME, name);
        element.setAttribute(Constants.ATTRIBUTE_TYPE, this.getType());
        return element;
    }

};