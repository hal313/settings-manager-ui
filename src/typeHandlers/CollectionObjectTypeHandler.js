import { isDefinedAndNotEmpty } from '../util/isDefinedAndNotEmpty.js';
import { createError } from '../util/createError.js';
import { Constants } from '../Constants.js';
import { isObjectArray } from '../util/isObjectArray.js';
import { getChildSettingElements } from '../util/getChildSettingElements.js';
import { getNameFromElement } from '../util/getNameFromElement.js';
import { getOneElement } from '../util/getOneElement.js';
import { getAllElements } from '../util/getAllElements.js';
import { isDefined } from '../util/isDefined.js';
import { SettingModifier } from '../util/SettingModifier.js';

const ATTRIBUTE_INDEX = 'data-setting-index';
const ATTRIBUTE_COLLECTION_VALUE_CONTAINER = 'data-setting-value-container';


/**
 * Gets the index from an element.
 *
 * @param {Element} element the element to get the index from
 * @returns {Number} the index assigned to the element, based on the attribute
 * @throws when the index is not present
 */
const getIndexFrom = (element) => {
    let index = element.getAttribute(ATTRIBUTE_INDEX);
    if (isDefined(index)) {
        return +index;
    }
    throw createError(`element does not have an index assigned via ${ATTRIBUTE_INDEX} "${element.outerHTML}"`);
};

/**
 * Gets an array item container element by index.
 *
 * @param {Element} rootElement the element to get the array item container element from
 * @param {Number} index the index of the array item container element to get
 * @returns {Element} the element marked as having the specified index
 */
const getArrayItemContainerElementByIndex = (rootElement, index) => {
    return getOneElement(`[${ATTRIBUTE_INDEX}="${index}"]`, rootElement);
};

/**
 * Gets an array item container element by index.
 *
 * @param {Element} rootElement the element to get the array item container element from
 * @param {Number} index the index of the array item container element to get
 * @returns {Element[]} the elements marked as having the specified index
 */
const getArrayItemContainersElementByIndex = (rootElement, index) => {
    return getAllElements(`[${ATTRIBUTE_INDEX}="${index}"]`, rootElement);
};

/**
 * Gets an array item container element
 *
 * @param {Element} rootElement the root element
 * @returns {Element[]} all elements which are root containers
 */
const getArrayItemContainerElements = (rootElement) => {
    return getAllElements(`[${ATTRIBUTE_COLLECTION_VALUE_CONTAINER}]`, rootElement);

}

/**
 * Creates an slement which is suitable for residing in a DOM and represents the array element value.
 *
 * @param {*} value the value to create an element for
 * @param {Number} index the index of the item
 * @param {SettingModifier} settingModifier the SettingModifer to assist
 * @returns {Element} the created element
 */
const createArrayItemContainerElement = (value, index, settingModifier) => {
    const name = `name_${index}`;
    const element = settingModifier.createElement(name, value, settingModifier);
    element.setAttribute(ATTRIBUTE_INDEX, index);
    element.setAttribute(ATTRIBUTE_COLLECTION_VALUE_CONTAINER, '');
    return element;
};

export class CollectionObjectTypeHandler {

    constructor() {}

    getType() {
        return 'collection:object';
    }

    getValue(element, settingModifier) {
        let childElements = getChildSettingElements(element);
        let valueArray = [];

        // For each child setting, add the name and value touple to the returned settings object
        childElements.forEach((childElement) => {
            // The value object represented by this child
            const valueObject = {};

            // The element may be
            const valueElement = getChildSettingElements(childElement)[0];
            // Get the name from the element
            const name = getNameFromElement(valueElement);
            // Get the value (which may be a recursive call)
            const value = settingModifier.getValue(valueElement);

            // Add the name and value to the value object
            valueObject[name] = value;

            // Push the object
            valueArray.push(valueObject);
        });

        return valueArray;
    }

    setValue(element, settings, settingModifier) {
        // TODO: Is this necessary?
        element.setAttribute(Constants.ATTRIBUTE_CONTAINER_ELEMENT, '');

        // Get the container elements?
        let arrayItemContainerElements = getArrayItemContainerElements(element);

        // Remove values from the DOM which do not exist in the value
        arrayItemContainerElements.forEach((arrayItemContainerElement) => {
            let index = getIndexFrom(arrayItemContainerElement);

            // If the index of the element is more than the length if the settings, remove it
            // because that means the new settings array is shorter than the existing DOM array
            if ((index + 1) > settings.length) {
                arrayItemContainerElement.remove();
            }
        });


        // TODO: Merge dom and settings so they stay in order based on index
        // Go through the settings array; if the nth element is in the dom, set the value; otherwise, append it at the end
        settings.forEach((setting, index) => {
            let arrayItemContainerElements = getArrayItemContainersElementByIndex(element, index);

            switch (arrayItemContainerElements.length) {
                case 0:
                    // Add a new DOM element
                    element.appendChild(createArrayItemContainerElement(setting, index, settingModifier));
                    break;
                case 1:
                    // Update an existing DOM element
                    let arrayItemContainerElement = getArrayItemContainerElementByIndex(element, index);
                    settingModifier.setValue(arrayItemContainerElement, setting);
                    break;
                default:
                    throw createError(`Duplicate elements in setting "${getNameFromElement(element)}"`);
            }
        });

        element.value = settings;
    }

    createElement(name, value, settingModifier) {
        if (!isDefinedAndNotEmpty(name)) {
            throw createError('The "name" parameter must be specified');
        }
        if (!isObjectArray(value)) {
            throw createError('The "value" parameter must be an array');
        }

        // Create the root element
        const rootElement = document.createElement('div');

        // Set the value
        rootElement.value = value;

        // Set attributes
        rootElement.setAttribute('name', name);
        rootElement.setAttribute(Constants.ATTRIBUTE_NAME, name);
        rootElement.setAttribute(Constants.ATTRIBUTE_CONTAINER_ELEMENT, '');
        rootElement.setAttribute(Constants.ATTRIBUTE_TYPE, this.getType());

        // Set the value
        this.setValue(rootElement, value, settingModifier);

        return rootElement;
    }

};
