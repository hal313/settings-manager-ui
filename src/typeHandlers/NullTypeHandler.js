import { isDefinedAndNotEmpty } from '../util/isDefinedAndNotEmpty.js';
import { createError } from '../util/createError.js';
import { Constants } from '../Constants.js';

export class NullTypeHandler {

    constructor() {}

    getType() {
        return 'null';
    }

    getValue(/*element*/) {
        return null;
    }

    setValue(element/*, value*/) {
        element.value = null;
    }

    createElement(name/* , value */) {
        if (!isDefinedAndNotEmpty(name)) {
            throw createError('The "name" parameter must be specified');
        }

        let element = document.createElement('div');
        element.setAttribute('name', name);
        element.setAttribute('value', null);

        this.setValue(element, null);

        return element;
    }

};