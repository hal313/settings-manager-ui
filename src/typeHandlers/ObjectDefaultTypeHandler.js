import { isDefined } from '../util/isDefined.js';
import { createError } from '../util/createError.js';
import { Constants } from '../Constants.js';
import { getChildSettingElements } from '../util/getChildSettingElements.js';
import { getNameFromElement } from '../util/getNameFromElement.js';

// TODO: Template for adding new child nodes (use a template, if present, for all? (${name}, ${type}, [${value}]))
export class ObjectDefaultTypeHandler {

    constructor() {}

    getType() {
        return 'object:default';
    }

    getValue(element, settingModifier) {
        let childElements = getChildSettingElements(element);
        let valueObject = {};

        // For each child setting, add the name and value touple to the returned settings object
        childElements.forEach((childElement) => {
            // Get the name from the element
            let name = getNameFromElement(childElement);
            // Get the value (which may be a recursive call)
            let value = settingModifier.getValue(childElement);

            // Add the name and value to the value object
            valueObject[name] = value;
        });

        return valueObject;
    }

    setValue(element, settings, settingModifier) {
        // Append the name to the root path
        let childElements = getChildSettingElements(element);
        let childElementNames = childElements.map(getNameFromElement);

        // Remove values from the DOM which do not exist in the value
        childElements.forEach((childElement) => {
            let name = getNameFromElement(childElement);

            // If the DOM has a value which is not in the object,
            if (isDefined(settings)) {
                if (!Object.keys(settings).includes(name)) {
                    childElement.remove();
                }
            }
        });


        // Update values from the values which exist in the DOM
        childElements.forEach((childElement) => {
            let name = getNameFromElement(childElement);

            // If the DOM has a value which is in the settings object
            if (childElementNames.includes(name)) {
                if (isDefined(settings)) {
                    settingModifier.setValue(childElement, settings[name]);
                }
            }
        });


        // Add new child nodes for values which do not exist in the DOM
        if (isDefined(settings)) {
            Object.keys(settings).forEach((name) => {
                let settingValue = settings[name];

                if (!childElementNames.includes(name)) {
                    // Create the element
                    let childElement = settingModifier.createElement(name, settingValue);

                    // Add to the element
                    element.append(childElement);
                }

            });
        }

        element.value = settings;
    }

    createElement(name, value, settingModifier) {
        if (!isDefined(name)) {
            throw createError('The "name" parameter must be specified');
        }

        // Create an element
        const element = document.createElement('div');

        // Set attributes
        element.setAttribute('name', name);
        // Set custom attributes
        element.setAttribute(Constants.ATTRIBUTE_CONTAINER_ELEMENT, '');

        // Set the value
        this.setValue(element, value, settingModifier);

        return element;
    }

};