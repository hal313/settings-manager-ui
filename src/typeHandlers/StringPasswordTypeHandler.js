import {isDefinedAndNotEmpty} from '../util/isDefinedAndNotEmpty.js';
import {createError} from '../util/createError.js';
import { Constants } from '../Constants.js';

export class StringPasswordTypeHandler {

    constructor() {}

    getType() {
        return 'string:password';
    }

    getValue(element) {
        return element.value;
    }

    setValue(element, value) {
        element.value = value;
    }

    createElement(name, value) {
        if (!isDefinedAndNotEmpty(name)) {
            throw createError('The "name" parameter must be specified');
        }

        let element = document.createElement('input');
        element.setAttribute('name', name);
        element.setAttribute('type', 'password');

        this.setValue(element, value);

        return element;
    }

};