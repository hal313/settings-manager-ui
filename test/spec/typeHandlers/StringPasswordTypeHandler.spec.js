import { StringPasswordTypeHandler } from '../../../src/typeHandlers/StringPasswordTypeHandler';
import { Constants } from '../../../src/Constants';

// TODO: Test types (or cast to boolean)
// TODO: Handle bad cases (null element, etc)
describe('StringPasswordTypeHandler', () => {

    describe('Lifecycle', () => {

        test('should exist', () => {
            expect(StringPasswordTypeHandler).toBeDefined();
        });

    });

    describe('API', () => {
        let stringPasswordTypeHandler;

        beforeEach(() => {
            stringPasswordTypeHandler = new StringPasswordTypeHandler();
        });

        describe('createElement', () => {

            test('should throw when name is undefined', () => {
                expect(() => stringPasswordTypeHandler.createElement(undefined, 'someValue')).toThrowError();
            });

            test('should throw when name is null', () => {
                expect(() => stringPasswordTypeHandler.createElement(null, 'someValue')).toThrowError();
            });

            test('should throw when name is the empty string', () => {
                expect(() => stringPasswordTypeHandler.createElement('', 'someValue')).toThrowError();
            });

            test('should throw when name is whitespace', () => {
                expect(() => stringPasswordTypeHandler.createElement('   ', 'someValue')).toThrowError();
            });

            test('should create a text string input element', () => {
                const name = 'fieldName';
                const value = 'some string value;'

                // Create the element
                const element = stringPasswordTypeHandler.createElement(name, value);

                // Check the type
                expect(element instanceof Element).toBeTruthy();
                // Check the required attributes
                expect(element.getAttribute('name')).toEqual(name);
                expect(element.getAttribute('type')).toEqual('password');
                // Check the value
                expect(element.value).toEqual(value);

                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();
            });

        });

        describe('getValue', () => {

            test('should get the value from a self-created element', () => {
                const name = 'fieldName';
                const value = 'some string value';

                // Create the element
                const element = stringPasswordTypeHandler.createElement(name, value);

                // Check the value
                expect(stringPasswordTypeHandler.getValue(element)).toEqual(value);
            });

        });

        describe('setValue', () => {

            test('should set the value for a self-created element', () => {
                const name = 'fieldName';
                const value = 'some string value';

                // Create the element
                const element = stringPasswordTypeHandler.createElement(name, value);

                // Precondition check
                expect(stringPasswordTypeHandler.getValue(element)).toEqual(value);
                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();

                // Set the new value
                const newValue = value.split('').reverse().join('');
                stringPasswordTypeHandler.setValue(element, newValue);

                // Check the new value
                expect(stringPasswordTypeHandler.getValue(element)).toEqual(newValue);

                // Check the value
                expect(element.value).toEqual(newValue);

                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();
            });

        });

    });

});