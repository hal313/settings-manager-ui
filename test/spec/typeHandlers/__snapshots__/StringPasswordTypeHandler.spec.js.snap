// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StringPasswordTypeHandler API createElement should create a text string input element 1`] = `

<input name="fieldName"
       type="password"
>

`;

exports[`StringPasswordTypeHandler API setValue should set the value for a self-created element 1`] = `

<input name="fieldName"
       type="password"
>

`;

exports[`StringPasswordTypeHandler API setValue should set the value for a self-created element 2`] = `

<input name="fieldName"
       type="password"
>

`;
